### PR TITLE
Projecttimeline expansion++

### PR DESCRIPTION
**Pull request by Remi749**
### Your checklist for this pull request

- [X] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). Don't request your main!
- [X] Make sure you are making a pull request against the **dev** branch (left side). Also you should start *your branch* off *dev*.
- [X] Check the commit's or even all commits' message 
- [X] Check if your code additions will fail linting checks
- [X] Remember: Add PR description to [CHANGELOG](https://github.com/Puzzlepart/prosjektportalen365/blob/dev/CHANGELOG.md) with the ID that matches this PR

### Description

This pull request is for adding the new Projecttimeline portoflio webpart. More information here: #434

### How to test
Moved to #434. Will merge to `dev` so CI adds this, then test.

### Relevant issues (if applicable)
#434

### CODE REVIEW

**Comment by olemp** [comment]
boom! 

image

File: SharePointFramework/PortfolioWebParts/src/components/ProjectTimeline/DetailsCallout/index.tsx [hunk 1]
```diff
@@ -11,28 +11,51 @@ export interface IDetailsCalloutProps {
...
   return (
     <Callout
       className={styles.detailsCallout}
       gapSpace={10}
       target={item.element}
       onDismiss={onDismiss}
       setInitialFocus={true}>
-      <p>
+      <p hidden={item.data.type !== 'Milepæl'}>
+        <b>Milepæl:</b>{' '}
```
Comment (olemp) [inline SharePointFramework/PortfolioWebParts/src/components/ProjectTimeline/DetailsCallout/index.tsx:24]: You're adding these as resources/strings, right? We don't want to add even more hardcoded strings 🚀

Comment (Remi749) [reply SharePointFramework/PortfolioWebParts/src/components/ProjectTimeline/DetailsCallout/index.tsx:24]: Of course ;)

Comment (olemp) [reply SharePointFramework/PortfolioWebParts/src/components/ProjectTimeline/DetailsCallout/index.tsx:24]: Better safe than sorry! Resolving convo!

**Comment by Remi749** [comment]
PR is ready for review

**Comment by olemp** [comment]
> PR is ready for review

[Build is failing](https://github.com/Puzzlepart/prosjektportalen365/pull/437/checks?check_run_id=2221487526)

image

**Comment by Remi749** [comment]
> > PR is ready for review
> 
> [Build is failing](https://github.com/Puzzlepart/prosjektportalen365/pull/437/checks?check_run_id=2221487526)
> 
> image

I'll look into it 👍

**Comment by Remi749** [comment]
@olemp, not sure if the spfx-package failed due to some of my changes done to the code as this is error in ProjectExtensions.
image

**Comment by olemp** [comment]
> @olemp, not sure if the spfx-package failed due to some of my changes done to the code as this is an error in ProjectExtensions.
> image

Ok, anyways, it needs fixing. I'll look into it.

File: SharePointFramework/PortfolioWebParts/src/components/ProjectTimeline/DetailsCallout/index.tsx [hunk 2]
```diff
@@ -18,21 +18,46 @@ export const DetailsCallout = ({ item, onDismiss }: IDetailsCalloutProps) => {
...
+      </p>
+      <p hidden={!item.data.projectUrl}>
         <b>Prosjekt:</b>{' '}
         <a href={item.data.projectUrl}>
           <span>{item.data.project}</span>
         </a>
       </p>
-      <p>
+      <p hidden={!item.data.phase}>
         <b>Gjeldende fase:</b> <span>{item.data.phase}</span>
```
Comment (olemp) [inline SharePointFramework/PortfolioWebParts/src/components/ProjectTimeline/DetailsCallout/index.tsx:36]: Add resource for this

File: SharePointFramework/PortfolioWebParts/src/components/ProjectTimeline/DetailsCallout/index.tsx [hunk 3]
```diff
@@ -18,21 +18,46 @@ export const DetailsCallout = ({ item, onDismiss }: IDetailsCalloutProps) => {
...
       </p>
-      <p>
+      <p hidden={item.data.type !== strings.MilestoneLabel}>
+        <b>Milepælsdato:</b> <span>{formatDate(item.data.end_time.toString())}</span>
```
Comment (olemp) [inline SharePointFramework/PortfolioWebParts/src/components/ProjectTimeline/DetailsCallout/index.tsx:39]: Add resource for this

File: SharePointFramework/PortfolioWebParts/src/components/ProjectTimeline/DetailsCallout/index.tsx [hunk 4]
```diff
@@ -18,21 +18,46 @@ export const DetailsCallout = ({ item, onDismiss }: IDetailsCalloutProps) => {
...
+      </p>
+      <p hidden={item.data.type === strings.MilestoneLabel}>
         <b>Startdato:</b> <span>{formatDate(item.data.start_time.toString())}</span>
```
Comment (olemp) [inline SharePointFramework/PortfolioWebParts/src/components/ProjectTimeline/DetailsCallout/index.tsx:42]: Add resource for this

File: SharePointFramework/PortfolioWebParts/src/components/ProjectTimeline/DetailsCallout/index.tsx [hunk 5]
```diff
@@ -18,21 +18,46 @@ export const DetailsCallout = ({ item, onDismiss }: IDetailsCalloutProps) => {
...
       </p>
-      <p>
+      <p hidden={item.data.type === strings.MilestoneLabel}>
         <b>Sluttdato:</b> <span>{formatDate(item.data.end_time.toString())}</span>
```
Comment (olemp) [inline SharePointFramework/PortfolioWebParts/src/components/ProjectTimeline/DetailsCallout/index.tsx:45]: Add resource for this

File: SharePointFramework/PortfolioWebParts/src/components/ProjectTimeline/DetailsCallout/index.tsx [hunk 6]
```diff
@@ -18,21 +18,46 @@ export const DetailsCallout = ({ item, onDismiss }: IDetailsCalloutProps) => {
...
       </p>
+      <p hidden={!item.data.budgetTotal}>
+        <b>Totalbudsjett:</b> <span>{tryParseCurrency(item.data.budgetTotal, '').toString().replace(/(?!^)(?=(?:\d{3})+(?:\.|$))/gm, ' ')}</span>
+      </p>
+      <p hidden={!item.data.costsTotal}>
+        <b>Kostnader påløpt totalt:</b> <span>{tryParseCurrency(item.data.costsTotal, '').toString().replace(/(?!^)(?=(?:\d{3})+(?:\.|$))/gm, ' ')}</span>
```
Comment (olemp) [inline SharePointFramework/PortfolioWebParts/src/components/ProjectTimeline/DetailsCallout/index.tsx:51]: Add resource for this

File: SharePointFramework/PortfolioWebParts/src/components/ProjectTimeline/DetailsCallout/index.tsx [hunk 7]
```diff
@@ -18,21 +18,46 @@ export const DetailsCallout = ({ item, onDismiss }: IDetailsCalloutProps) => {
...
+      </p>
+      <p hidden={item.data.type === strings.MilestoneLabel}>
         <b>Startdato:</b> <span>{formatDate(item.data.start_time.toString())}</span>
       </p>
-      <p>
+      <p hidden={item.data.type === strings.MilestoneLabel}>
         <b>Sluttdato:</b> <span>{formatDate(item.data.end_time.toString())}</span>
       </p>
+      <p hidden={!item.data.budgetTotal}>
+        <b>Totalbudsjett:</b> <span>{tryParseCurrency(item.data.budgetTotal, '').toString().replace(/(?!^)(?=(?:\d{3})+(?:\.|$))/gm, ' ')}</span>
```
Comment (olemp) [inline SharePointFramework/PortfolioWebParts/src/components/ProjectTimeline/DetailsCallout/index.tsx:48]: Add resource for this

File: SharePointFramework/PortfolioWebParts/src/components/ProjectTimeline/DetailsCallout/index.tsx [hunk 8]
```diff
@@ -18,21 +18,46 @@ export const DetailsCallout = ({ item, onDismiss }: IDetailsCalloutProps) => {
...
+      <p hidden={!item.data.costsTotal}>
+        <b>Kostnader påløpt totalt:</b> <span>{tryParseCurrency(item.data.costsTotal, '').toString().replace(/(?!^)(?=(?:\d{3})+(?:\.|$))/gm, ' ')}</span>
+      </p>
+      <p hidden={(!item.data.budgetTotal || !item.data.costsTotal) || item.data.type !== strings.ProjectLabel}>
+        <a target='_blank' rel='noreferrer' href={`${item.data.projectUrl}/SitePages/Prosjektstatus.aspx`}>
+          <span>{strings.LastPublishedStatusreport}</span>
+        </a>
+      </p>
+      <p hidden={!item.data.type || item.data.type === strings.ProjectLabel}>
+        <b>Type:</b> <span>{item.data.type}</span>
```
Comment (olemp) [inline SharePointFramework/PortfolioWebParts/src/components/ProjectTimeline/DetailsCallout/index.tsx:59]: Add resource for this

File: SharePointFramework/PortfolioWebParts/src/components/ProjectTimeline/index.tsx [hunk 9]
```diff
@@ -289,10 +320,48 @@ export class ProjectTimeline extends Component<IProjectTimelineProps, IProjectTi
...
         projectUrl: project.url,
-        phase: project.phase
+        phase: project.phase,
+        type: project.type,
+        budgetTotal: project.budgetTotal,
+        costsTotal: project.costsTotal
+      } as ITimelineItem
+    })
+
+    const phases: ITimelineItem[] = timelineItems.filter((item) => item.type !== strings.ProjectLabel).map((item, id) => {
```
Comment (olemp) [inline SharePointFramework/PortfolioWebParts/src/components/ProjectTimeline/index.tsx:330]: What's the filter doing?

File: SharePointFramework/PortfolioWebParts/src/components/ProjectTimeline/index.tsx [hunk 10]
```diff
@@ -289,10 +320,48 @@ export class ProjectTimeline extends Component<IProjectTimelineProps, IProjectTi
...
+
+      const backgroundColor = item.type === strings.PhaseLabel
+        ? '#2589d6'
+        : item.type === strings.MilestoneLabel
+          ? 'transparent'
+          : item.type === strings.SubPhaseLabel
+            ? '#249ea0'
+            : '#484848'
+
+      const group = _.find(groups, (grp) => item.title.indexOf(grp.title) !== -1)
```
Comment (olemp) [inline SharePointFramework/PortfolioWebParts/src/components/ProjectTimeline/index.tsx:341]: This function seems a bit complex. Can anything be simplified? Open question <3

**Review by olemp — CHANGES_REQUESTED** [review]

File: SharePointFramework/PortfolioWebParts/src/components/ProjectTimeline/index.tsx (hunk 9 above)
Comment (Remi749) [reply SharePointFramework/PortfolioWebParts/src/components/ProjectTimeline/index.tsx:330]: I see that this is no longer needed, will remove it. Thanks for noticing!

File: SharePointFramework/PortfolioWebParts/src/components/ProjectTimeline/index.tsx (hunk 10 above)
Comment (Remi749) [reply SharePointFramework/PortfolioWebParts/src/components/ProjectTimeline/index.tsx:341]: Best I can think of as of now would be:
`const group = _.find(groups, (grp) => project.title.includes(grp.title))`

Not sure if we should start using ".includes" as it doesn't have Internet Explorer support though. ([source](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/includes#browser_compatibility))

Comment (olemp) [reply SharePointFramework/PortfolioWebParts/src/components/ProjectTimeline/index.tsx:341]: Underscore has `contains` 🚀

**Pull request merged**